src/rewrite_gelu.cpp

```diff
@@ -33,74 +33,75 @@ namespace migraphx {
-        auto sig = m.insert_instruction(ins, make_op("neg"), mul);
-        sig      = m.insert_instruction(ins, make_op("exp"), sig);
-        auto one = m.add_literal(literal{shape{x->get_shape().type()}, {1.0f}});
-        sig      = insert_common_op(m, ins, make_op("add"), {sig, one});
-        sig      = m.insert_instruction(ins, make_op("div"), x, sig);
-        m.replace_instruction(ins, sig);
+        replace_with_tanh_exp_gelu(m, r);
     }
 };
 
 /**
- * Find fastGELU blocks (where the graph already does a GELU approximation) and replace them
- * with an alternative approximation that is less likely to overflow.
- * The replacement approximation is equivalent to:
- * GELU(x) ~= 0.5 * x * ( 1 + tanh( sqrt(2/M_PI) * (x + 0.044715 * x^3)))
- * You can rearrange to the form used in this by recognizing that
- * 1 + tanh(x) = (2) / (1 + exp(-2 * x)).
- * The fitting constant 0.044715 is from
- * A. Choudhury, ‘A simple approximation to the area under standard normal curve’, Mathematics and
- * Statistics, vol. 2, no. 3, pp. 147–149, 2014.
+ * Find tanhGELU blocks and replace them with a rearranged version that is less likely to overflow
+ * and is more performant.
  */
 struct find_tanh_fast_gelu
 {
     auto matcher() const { return match::gelu_tanh(); }
 
     void apply(module& m, const match::matcher_result& r) const
     {
-        auto ins            = r.result;
-        auto x              = r.instructions["x"];
-        double sqrt_2_rpi   = sqrt(M_2_PI);
-        auto sqrt_2_rpi_lit = m.add_literal(literal{shape{x->get_shape().type()}, {sqrt_2_rpi}});
-        auto fit_const      = m.add_literal(literal{shape{x->get_shape().type()}, {0.044715f}});
-        auto one            = m.add_literal(literal{shape{x->get_shape().type()}, {1.0f}});
-        auto xb             = insert_common_op(m, ins, make_op("mul"), {x, sqrt_2_rpi_lit});
-        auto a              = insert_common_op(m, ins, make_op("mul"), {xb, fit_const});
-        auto b              = m.insert_instruction(ins, make_op("mul"), a, x);
-        auto c              = m.insert_instruction(ins, make_op("mul"), b, x);
-        auto u              = m.insert_instruction(ins, make_op("add"), c, xb);
-        auto neg_u          = m.insert_instruction(ins, make_op("neg"), u);
-        auto d              = m.insert_instruction(ins, make_op("sub"), neg_u, u);
-        auto emu            = m.insert_instruction(ins, make_op("exp"), d);
-        auto e              = insert_common_op(m, ins, make_op("add"), {one, emu});
-        auto cdf            = insert_common_op(m, ins, make_op("div"), {one, e});
-        auto y              = m.insert_instruction(ins, make_op("mul"), x, cdf);
-        m.replace_instruction(ins, y);
+        replace_with_tanh_exp_gelu(m, r);
     }
 };
 
 void rewrite_gelu::apply(module& m) const
 {
     if(fast_math)
     {
-        match::find_matches(m, find_gelu_erf{}, find_tanh_fast_gelu{});
+        match::find_matches(m, find_gelu_erf{});
```

Comment:
BTW, if we match fast_math clause, I would expect tanh_fast_gelu to be used..  I am getting mixed up here.. sorry.

Reply:
We have a flag `--disable-fast-math` that's for preventing replacing erf -> tanh_exp approx. `find_gelu_erf` by itself already matches both erf and the slower tanh approximations.

Reply:
My expectation would be that `disable-fast-math` should logically turn off the `fast_math` variable here.. so taking out the `find_tanh_fast_gelu()` from this matcher clause is very confusing to me, when it is listed under `if(fast_math)`.

Reply:
I agree it's somewhat confusing. I changed L70 in rewrite_gelu since it was requested during an earlier meeting:
> auto matcher() const { return match::any_of(match::gelu_erf(), match::gelu_tanh()); }
We still want to replace the slower tanh approximations if found, because we don't lose any accuracy.